src/partial-result-stream.ts

```diff
@@ -187,20 +196,57 @@ export class PartialResultStream extends Transform implements ResultEvents {
         .fields as google.spanner.v1.StructType.Field[];
     }
 
+    let res = true;
     if (!is.empty(chunk.values)) {
-      this._addChunk(chunk);
+      res = this._addChunk(chunk);
     }
 
-    next();
+    if (res || !this._requestStream) {
+      next();
+    } else {
+      // Wait a little before we push any more data into the pipeline as a
+      // component downstream has indicated that a break is needed. Pause the
+      // request stream to prevent it from filling up the buffer while we are
+      // waiting.
+      setTimeout(() => {
+        this._tryResume(next, 4);
+      }, 2);
+    }
+  }
+
+  private _tryResume(next: Function, timeout: number) {
+    // Try to push an empty chunk to check whether more data can be accepted.
+    if (this.push(undefined)) {
+      this._numPushFailed = 0;
+      this._requestStream!.resume();
+      next();
+    } else {
+      // Downstream returned false indicating that it is still not ready for
+      // more data.
+      this._numPushFailed++;
+      setTimeout(() => {
+        const nextTimeout = Math.min(timeout * 2, 1024);
+        if (this._numPushFailed === this._options.maxResumeRetries) {
```

Comment:
We might as well bump this out of the setTimeout to fail as quick as possible when we know we're done trying.

Reply:
Erm... Yeah, that was quite useless to have it inside the `setTimeout`. I've moved it out.